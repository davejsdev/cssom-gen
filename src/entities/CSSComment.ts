export type CSSCommentInlineCSSText = `/* ${string} */`;
export type CSSCommentBlockCSSText = `/**\n${string}\n */`;
export type CSSCommentCSSText = CSSCommentInlineCSSText | CSSCommentBlockCSSText;

export type CSSCommentType = 'inline' | 'block';
export interface CSSCommentOptions {
  type: CSSCommentType;
}

/**
 * A comment in a CSS stylesheet.
 */
export class CSSComment {
  static defaultOptions: CSSCommentOptions = {
    type: 'inline',
  };

  value: string;
  type: CSSCommentType;

  constructor(value: string, options: CSSCommentOptions = CSSComment.defaultOptions) {
    this.value = value;
    this.type = options.type;
  }

  get cssText(): CSSCommentCSSText {
    if (this.type === 'inline') {
      return `/* ${this.value} */`;
    }

    if (this.type === 'block') {
      // Format the value as a block comment
      const lines = this.value.split('\n');
      const indentedLines = lines.map((line) => ` * ${line}`);
      const formattedLines = indentedLines.join('\n');

      return `/**\n${formattedLines}\n */`;
    }

    throw new Error('Invalid comment type');
  }
}
