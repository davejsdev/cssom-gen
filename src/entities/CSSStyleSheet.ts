// biome-ignore lint/suspicious/noTsIgnore: no types available for this package
// @ts-ignore-next-line no-restricted-globals
import Window from 'window';
import type { CSSComment } from './CSSComment';

const window = new Window();

interface CSSStyleSheetInit extends Window.CSSStyleSheetInit {
  nodes: CSSComment[];
}

export class CSSStyleSheet extends window.CSSStyleSheet {
  nodes: CSSComment[];

  constructor(options: CSSStyleSheetInit) {
    super(options);

    this.nodes = options.nodes || [];
  }

  append(node: CSSComment) {
    this.nodes.push(node);
  }

  remove(node: CSSComment) {
    this.nodes = this.nodes.filter((n) => n !== node);
  }

  get cssText() {
    return `${this.nodes.map((node) => node.cssText).join('\n\n')}\n`;
  }
}
