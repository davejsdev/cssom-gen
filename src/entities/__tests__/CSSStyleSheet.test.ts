import { describe, expect, it } from 'vitest';
import { CSSComment } from '../CSSComment';
import { CSSStyleSheet } from '../CSSStyleSheet';

describe('CSSStyleSheet', () => {
  it('should create a stylesheet', () => {
    const stylesheet = new CSSStyleSheet({ nodes: [] });
    expect(stylesheet.cssText).toBe('\n');
  });

  it.each([[[]], [[new CSSComment('This is a block comment', { type: 'block' })]]])(
    'always ends with a single blank newline',
    (nodes) => {
      const stylesheet = new CSSStyleSheet({ nodes });
      expect(stylesheet.cssText.endsWith('\n')).toBe(true);
      expect(stylesheet.cssText.endsWith('\n\n')).toBe(false);
    },
  );

  it('should create a stylesheet with a comment', () => {
    const stylesheet = new CSSStyleSheet({ nodes: [] });
    const comment = new CSSComment('This is a block comment', { type: 'block' });
    stylesheet.append(comment);
    expect(stylesheet.cssText).toMatchSnapshot();
  });
});
