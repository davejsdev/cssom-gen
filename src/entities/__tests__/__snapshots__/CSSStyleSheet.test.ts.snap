// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CSSStyleSheet > should create a stylesheet with a comment 1`] = `
"/**
 * This is a block comment
 */
"
`;
